File: python_cpp/uPIMulator_frontend/src/abi/isa/cc/ext_sub_set_cc.py
from typing import Set

from abi.isa.cc._base_cc import BaseCC
from abi.isa.instruction.condition import Condition


class ExtSubSetCC(BaseCC):
    def __init__(self, condition: Condition):
        super().__init__(
            ExtSubSetCC.conditions(), condition,
        )

    @staticmethod
    def conditions() -> Set[Condition]:
        return {
            Condition.C,
            Condition.NC,
            Condition.Z,
            Condition.NZ,
            Condition.XZ,
            Condition.XNZ,
            Condition.OV,
            Condition.NOV,
            Condition.EQ,
            Condition.NEQ,
            Condition.PL,
            Condition.MI,
            Condition.SZ,
            Condition.SNZ,
            Condition.SPL,
            Condition.SMI,
            Condition.GES,
            Condition.GEU,
            Condition.GTS,
            Condition.GTU,
            Condition.LES,
            Condition.LEU,
            Condition.LTS,
            Condition.LTU,
            Condition.XGTS,
            Condition.XGTU,
            Condition.XLES,
            Condition.XLEU,
            Condition.TRUE,
        }

File: golang/uPIMulator/src/linker/kernel/instruction/cc/ext_sub_set_cc.go
package cc

import (
	"errors"
)

type ExtSubSetCc struct {
	condition Condition
}

func (this *ExtSubSetCc) Init(condition Condition) {
	conditions := map[Condition]bool{
		C:    true,
		NC:   true,
		Z:    true,
		NZ:   true,
		XZ:   true,
		XNZ:  true,
		OV:   true,
		NOV:  true,
		EQ:   true,
		NEQ:  true,
		PL:   true,
		MI:   true,
		SZ:   true,
		SNZ:  true,
		SPL:  true,
		SMI:  true,
		GES:  true,
		GEU:  true,
		GTS:  true,
		GTU:  true,
		LES:  true,
		LEU:  true,
		LTS:  true,
		LTU:  true,
		XGTS: true,
		XGTU: true,
		XLES: true,
		XLEU: true,
		TRUE: true,
	}

	if _, found := conditions[condition]; !found {
		err := errors.New("condition is not allowed")
		panic(err)
	}

	this.condition = condition
}

func (this *ExtSubSetCc) Condition() Condition {
	return this.condition
}
